Added Models and Controllers

ZachCarrillo opened the pull request:
Added Customer, CustomerGroup, and GroupMember models and controllers for each

ZachCarrillo committed:
Added Models and Controllers

Added Customer, CustomerGroup, and GroupMember models and controllers for each

ysemerikov commented on .gitignore:
```diff
@@ -348,3 +348,4 @@ MigrationBackup/
 
 # Ionide (cross platform F# VS Code tools) working folder
 .ionide/
+CustomerContactBook/CustomerContactBook/obj/project.assets.json
```
Feel free to copy `.gitignore` from [my repository](https://github.com/ysemerikov/oa-customer-contact-book/blob/main/.gitignore)

ysemerikov commented on CustomerContactBook/CustomerContactBook/Controllers/CustomerGroupsController.cs:
```diff
@@ -0,0 +1,107 @@
+                {
+                    throw;
+                }
+            }
```
```suggestion
            catch (DbUpdateConcurrencyException) when (!CustomerGroupExists(id))
            {
                return NotFound();
            }
```

ZachCarrillo replied:
I have made this charge to each controller and pushed it to this branch

ysemerikov commented on CustomerContactBook/CustomerContactBook/Controllers/CustomerGroupsController.cs:
```diff
@@ -0,0 +1,107 @@
+
+        private bool CustomerGroupExists(int id)
+        {
+            return _context.Groups.Any(e => e.Id == id);
```
I see in most places you use `Async` version of methods to access database, why not here?

ysemerikov commented on CustomerContactBook/CustomerContactBook/Models/Customer.cs:
```diff
@@ -0,0 +1,20 @@
+            {
+                return LastName + ", " + FirstName;
+            }
+        }
```
```suggestion
        public string FullName => LastName + ", " + FirstName;
```

ysemerikov commented on CustomerContactBook/CustomerContactBook/Models/Customer.cs:
```diff
@@ -0,0 +1,20 @@
+        public string? PhoneNum { get; set; }
+        public string? Email { get; set; }
+        public string? Address { get; set; }
+        public DateTime BirthDay { get; set; }
```
```suggestion
        public DateTime? BirthDay { get; set; }
```

ysemerikov commented on CustomerContactBook/CustomerContactBook/Models/Customer.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public class Customer
+    {
+        public int Id { get; set; }
```
`int.MaxValue` is small for database identifier, please use `long`

ysemerikov commented on CustomerContactBook/CustomerContactBook/Controllers/CustomerGroupsController.cs:
```diff
@@ -0,0 +1,107 @@
+    [ApiController]
+    public class CustomerGroupsController : ControllerBase
+    {
+        private readonly CustomerGroupContext _context;
```
FYI, DatabaseContext shouldn't be used in the Controller level, there should be CustomerService.
Anyway it's fine for now, at the next steps we will allocate service layer.

ysemerikov commented on CustomerContactBook/CustomerContactBook/Controllers/CustomerGroupsController.cs:
```diff
@@ -0,0 +1,107 @@
+        // PUT: api/CustomerGroups/5
+        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
+        [HttpPut("{id}")]
+        public async Task<IActionResult> PutCustomerGroup(int id, CustomerGroup customerGroup)
```
`CustomerGroup` is class that directly mapped into database table, for controllers please use another one (`CustomerGroupDto`), like described in [the link](https://go.microsoft.com/fwlink/?linkid=2123754) from comment of this method.

So there will be:
```
public class CustomerGroup // this class maps to database
{
    public long Id { get; set; }
    public string Name { get; set; }
}

public class CustomerGroupDto // this class is parameter of controller methods
{
    public long Id { get; set; }
    public string Name { get; set; }
}

public class CustomerGroupCreateDto // this class is parameter of controller methods
{
    public string Name { get; set; }
}
```

ysemerikov reviewed (commented):
I glad you start to use EF to access database, I will tell more about this on the future sessions.

ZachCarrillo committed:
Update BirthDay in Customer Model

Declare birthday as nullable

Co-authored-by: Yury Semerikov <yury.semerikov@gmail.com>

ZachCarrillo committed:
Update FullName in Customer.cs

Co-authored-by: Yury Semerikov <yury.semerikov@gmail.com>

ZachCarrillo committed:
Add Suggested Changes

changed ids for models from int to long.
changed catch(dbUpdatreConcurrencyExeption) for each controller

ysemerikov reviewed (approved):
synced in the slack:
'DTO' classes will be introduced in the next PR
everything else is fine to merge now.
